clients/cobol-lsp-vscode-extension/src/services/copybook/downloader/ZoweExplorerDownloader.ts

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2024 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+import * as vscode from "vscode";
+import { DownloadUtil } from "./DownloadUtil";
+import * as iconv from "iconv-lite";
+import { SettingsService } from "../../Settings";
+import { CopybookURI } from "../CopybookURI";
+import {
+  DOWNLOAD_QUEUE_LOCKED_ERROR_MSG,
+  INVALID_CREDENTIALS_ERROR_MSG,
+  PROFILE_NAME_PLACEHOLDER,
+  UNLOCK_DOWNLOAD_QUEUE_MSG,
+} from "../../../constants";
+import { TelemetryService } from "../../reporter/TelemetryService";
+
+export abstract class ZoweExplorerDownloader {
+  private static lockedProfile: Set<string> = new Set();
+
+  constructor(
+    private readonly storagePath: string,
+    protected readonly explorerAPI: IApiRegisterClient | undefined,
+  ) {}
+
+  protected abstract downloadCopybookContent(
+    dataset: string,
+    member: string,
+    profileName: string,
+  ): Promise<void>;
+
+  protected getZoweProfile(profileName: string) {
+    return this.explorerAPI!.getExplorerExtenderApi()
+      .getProfilesCache()
+      .loadNamedProfile(profileName);
+  }
+
+  protected getDownloadOptions(
+    profileName: string,
+    dataset: string,
+    member: string,
+    loadedProfile: IProfileLoaded,
+  ) {
+    const downloadBinary = !!SettingsService.getCopybookFileEncoding();
+    const baseUri = vscode.Uri.file(
+      CopybookURI.createDatasetPath(profileName, dataset, this.storagePath),
+    );
+    const filePath = vscode.Uri.joinPath(baseUri, member);
+
+    return {
+      file: filePath,
+      binary: downloadBinary,
+      returnEtag: true,
+      ...(downloadBinary ? {} : { encoding: loadedProfile.profile.encoding }),
```

Comment:
I think the "manual" encoding is missing here. The original code either passed encoding to zowe or requested binary download and performed the conversion manually.

The current version either download binary file, or performs the conversion twice (once zowe and once manual).

Reply:
updated